text/3323-restrictions.md

```diff
@@ -0,0 +1,466 @@
+`pub(super)` to restrict visibility to the parent module, and `pub(in path)` to restrict visibility
+to `path`, as long as that path is an ancestor of the location it is used. There is one additional
+case you have likely never encountered: `pub(self)`. The reason you have likely never seen this
+before is that it is completely useless. Why? `pub(self)` is identical to private, which is the
+default visibility.
+
+While `pub(self)` should never be written in ordinary code, `impl(self)` and `mut(self)` are quite
+different. As the default restriction is unrestricted implementation and unrestricted mutation, it
+is more than likely that `impl(self)` and `mut(self)` will be quite common. During previous
+discussions about syntax, the unclear meaning of `impl(self)` and `mut(self)` was brought up. Syntax
+should have a clear meaning, and `impl(self)` and `mut(self)` are not clear. While they should be
+allowed for consistency with `pub(self)`, an alternative was needed with the same behavior.
+`impl(mod)` and `mut(mod)` should have a sufficiently clear meaning: implementing the trait and
+mutating the field are restricted to the current module. `impl(in mod)` and `mut(in mod)` are
+accepted as well. The behavior of `impl(mod)` is identical to `impl(self)`; likewise for `mut(mod)`
+and `mut(self)`.
+
+For consistency with the new restriction syntax, `pub(mod)` and `pub(in mod)` are also allowed.
+
+# Reference-level explanation
+
+## Syntax
+
+Using the syntax from [the reference for `struct`s][struct syntax], the change needed to support
+`mut` restrictions is quite small.
+
+[struct syntax]: https://doc.rust-lang.org/stable/reference/items/structs.html
+
+```diff
+StructField :
+   OuterAttribute*
+   Visibility?
++  MutRestriction?
+   IDENTIFIER : Type
+
+TupleField :
+   OuterAttribute*
+   Visibility?
++  MutRestriction?
+   Type
+
++MutRestriction :
++     mut
++   | mut ( crate )
++   | mut ( self )
++   | mut ( mod )
++   | mut ( super )
++   | mut ( in SimplePath )
++   | mut ( in mod )
+```
+
+Trait definitions need a similar change to the [syntax for `trait`s][trait syntax] to accommodate
+`impl` restrictions.
+
+[trait syntax]: https://doc.rust-lang.org/stable/reference/items/traits.html
+
+```diff
+Trait :
+   unsafe?
++  ImplRestriction?
+   trait IDENTIFIER
+   GenericParams? ( : TypeParamBounds? )? WhereClause? {
+     InnerAttribute*
+     AssociatedItem*
+   }
+
++ImplRestriction :
++     impl
++   | impl ( crate )
++   | impl ( self )
++   | impl ( mod )
++   | impl ( super )
++   | impl ( in SimplePath )
++   | impl ( in mod )
+```
+
+Essentially, `mut` and `impl` have the same syntax as `pub`, just with a different keyword.
+
+Finally, a minor change is needed to the syntax for visibility to permit `pub(mod)`.
+
+```diff
+Visibility :
+     pub
+   | pub ( crate )
+   | pub ( self )
++  | pub ( mod )
+   | pub ( super )
+   | pub ( in SimplePath )
++  | pub ( in mod )
+```
+
```

Comment:
```suggestion
Finally, a minor change is needed to the syntax for visibility to permit `pub(mod)`.

```diff
Visibility :
     pub
   | pub ( crate )
   | pub ( self )
+  | pub ( mod )
   | pub ( super )
   | pub ( in SimplePath )
+  | pub ( in mod )
```

I think if we want to consider doing this at all, it should be a completely separate RFC.
```